extensions/emmet/src/abbreviationActions.ts

```diff
@@ -616,14 +616,30 @@ function expandAbbr(input: ExpandAbbreviationInput): string | undefined {
 				}
 
 				// If wrapping with a block element, insert newline in the text to wrap.
-				if (wrappingNode && inlineElements.indexOf(wrappingNode.name) === -1) {
-					wrappingNode.value = '\n\t' + wrappingNode.value + '\n';
+				if (wrappingNode) {
+					if (inlineElements.indexOf(wrappingNode.name) === -1) {
+						wrappingNode.value = '\n\t' + wrappingNode.value + '\n';
+					}
+
+					// Fixes #54711
+					if (wrappingNode.name === 'a') {
+						let hrefAtt = wrappingNode.attributes.filter((a) => a.name === 'href')[0];
+						if (hrefAtt) {
+							hrefAtt.value = input.textToWrap[0]; // This is safe because we've already checked that length is 1.
```

Comment:
Shouldn't we do this only if the text to wrap is a url or an email?

Reply:
> Sure, but we don't have that information since the input.textToWrap is just $TM_SELECTED_TEXT

Good point. Perhaps, we can pass the actual text in the `input.textToWrap` to the `expandAbbr` function and move the https://github.com/Microsoft/vscode/blob/6ad61f018ab84d1982dd9dacdb6e7eefb1de8d33/extensions/emmet/src/abbreviationActions.ts#L193 logic to inside `expandAbbr` ?

> It should be a url or email 90% of the time anyway, no?

Most of the time, a link has a readable link text which is different from the url.

Reply:
I've had a play around with removing the logic in 
https://github.com/Microsoft/vscode/blob/6ad61f018ab84d1982dd9dacdb6e7eefb1de8d33/extensions/emmet/src/abbreviationActions.ts#L189-L194

With something like this;
``` typescript
let textToWrap = rangesAndContent.textToWrapInPreview;
```

I then have to determine *when* to use `$TM_SELECTED_TEXT` as `expandAbbr` is used by both `applyPreview` and `makeChanges` (a.k.a. final result). 

I think the easiest option would be to add a new parameter to `expandAbbr` that indicates whether or not this is being called by the preview code or not, which can default to false. Perhaps
`expandAbbr(input: ExpandAbbreviationInput, isPreview: boolean)` would suffice. Thoughts?

Reply:
> I think the easiest option would be to add a new parameter to expandAbbr that indicates whether or not this is being called by the preview code or not, which can default to false. 

Yes, that works for me.